lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+
+data PointerUTxO = PointerUTxO { pTxIn :: TxIn, pCoin :: Coin }
+    deriving (Generic, Eq, Show)
+    deriving anyclass NFData
+
+-- | Returns the index corresponding to the payment key the `PointerUTxO`
+-- should be locked with for a portfolio of a given size @n@.
+--
+-- In our current implementation we require the `PointerUTxO` to be created in
+-- the @[0] -> [0,1] transition@, i.e. @nextKeyIx 1 -> nextKeyIx 2@.
+pointerIx
+    :: Int
+    -> Maybe (Index 'Soft 'AddressK)
+pointerIx 0 = Nothing
+pointerIx 1 = Nothing
+pointerIx n = Just $ toEnum n
+
+--------------------------------------------------------------------------------
+-- Constructors
+--------------------------------------------------------------------------------
+
+-- | Construct the initial delegation state.
+initialDelegationState
+    :: k 'AccountK XPub
+    -> DelegationState k
+initialDelegationState accK = DelegationState accK Zero
+
+--------------------------------------------------------------------------------
+-- Chain
+--------------------------------------------------------------------------------
+
+-- | A transaction type specific to `DelegationState`.
+--
+-- Intented to be converted both from and to a more real transaction type.
+--
+-- When constructing a real transaction from `Tx`, these `outputs`
+-- should appear before other outputs. In the theoretical event that there's
+-- also a user-specified output with the same payment key as the pointer output,
+-- `applyTx` will track the first one as the new pointer.
+data Tx = Tx
+    { certs :: [Cert]
+    , inputs :: [(TxIn, Coin)]
+    , outputs :: [TxOut]
+    }
+    deriving (Eq, Generic)
+    deriving Show via (Quiet Tx)
+
+instance Semigroup Tx where
+    (Tx cs1 is1 os1) <> (Tx cs2 is2 os2) =
+        Tx (cs1 <> cs2) (is1 <> is2) (os1 <> os2)
+
+data Cert
+    = RegisterKey RewardAccount
+    | Delegate RewardAccount
+      -- ^ Which pool we're delegating to is here (and for now) irrelevant.
+      -- The main thing is that there exists a witness on-chain for this stake
+      -- key (registration certs don't require witnesses)
+      --
+      -- TODO: We may also want to add the PoolId here.
+    | DeRegisterKey RewardAccount
+    deriving (Eq, Show, Generic)
+
+-- | Given a `DelegationState`, produce a `Tx` registering or de-registering
+-- stake-keys, in order to have @n@ stake-keys.
+--
+-- E.g. @setPortfolioOf s0 _ _ 2@ creates a tx which after application causes
+-- the state to have @activeKeys == [0,1]@
+--
+-- Returns @Nothing@ if the target @n@ is already reached.
+setPortfolioOf
+    :: (SoftDerivation k, ToRewardAccount k)
+    => DelegationState k
+    -> Coin
+        -- ^ minUTxOVal
+    -> (k 'AddressK XPub -> Address)
+        -- ^ A way to construct an Address
+    -> (RewardAccount -> Bool)
+        -- ^ Whether or not the key is registered.
+        --
+        -- TODO: Need a Set or Map for the real implementation with LSQ.
```

Comment:
How are you thinking to approach this function without any monadic effects though :thinking:  ? The caller of `setPortfolioOf` won't know the accounts upfront, so there's little chance that it'll be able to resolve that predicate without having to query the node? Or am I missing something?

Reply:
I'm thinking something like:
```haskell
let accounts = Set.fromList $ accoutAtIx [0 .. n]
(registrationStatus :: Map RewardAccount Bool) <- checkIfRegistered nl
setPortfolioOf n registrationStatus
```
Or taking a function of `Set RewardAccount -> m (Map RewardAccount Bool)` (and making `setPortfolioOf n` return `m` too) if it turns out to be neater.